src/jsep.js

```diff
@@ -617,37 +647,46 @@ let jsep = function(expr) {
 		};
 	};
 
-	let nodes = [], ch_i, node;
+	const hookScope = {
```

Comment:
See `env` is supposed to be the local environment with properties only for the variables of the current scope, not an object to be reused like this. You've noticed that this approach is awkward, as you need to null out properties etc. This makes it very easy to introduce bugs by modifying this object, not realizing it's re-used across hooks.

I quite like this approach with the accessor on `index`, it's much more elegant than using `env.index` throughout, but in balance, I think the disadvantages of re-using the env object outweigh the advantages. You can of course reuse `env` when it's in the same scope.

For exposing the functions to plugins, we could just export them as named exports.

Reply:
Hmm... I'm curious if you have any suggestions? I was trying to avoid adding a property accessor for existing jsep code on all usages of the scoped variables (and the functions that rely on them). If it were all a class using accessors, then the entire class could be passed as the env. The node (and nodes) have to be on the stack, but the others are all more global. I think named exports would lose the scope of the current js expression parsing...

I was also trying to avoid regenerating that object with every call to the hooks which could slow it down more, something like this:
```javascript
hooks.run('name', { node });
function run(name, env) {
  hooks[name].forEach(h => h({ ...baseEnv, ...env }));
}
```

Any suggestions on which way to go?

Reply:
If we have very few hooks, it may be easier to just do `index = env.index` after the hook. Once we get more, we can start just using `env.index` instead of `index`. 

I think at this point creating object literals has become pretty fast, and avoiding it for perf was a micro-optimization we did a long time ago? But I might always be wrong! Are you aware of any recent benchmarks?

You are right regarding named exports for some of these functions, and exports have to be top level anyway. I wonder if it would make sense to pass the index and expression to them as a parameter, so they could be top level. That would also make it clear what their input is. 

Alternatively, we could go the direction you suggested and make them class methods, but `jsep()` would still need to work as a function and return the AST. Since ES6 constructors cannot be used as both (using a constructor as a function throws), it would need to be a separate `Parser` (name TBB) class that we can export as a named export, and `jsep` would still be the default export. It would create a `Parser` instance and return the AST, so that existing usage is the same. In this case, hooks would indeed just get `this` as their context.

What do you think?

Reply:
I ran a couple benchmarks on jsbench.me out of curiosity...

For the [env variable](https://jsbench.me/kcknfjp4zo/1), recreating the env per hook call is 50-280 times slower than sharing the env:
![image](https://user-images.githubusercontent.com/17912877/114498539-1eb02180-9bd9-11eb-8cff-71b95e2b6572.png)

However, I ran another test comparing direct variable access versus object property access (i.e. adding `this.` in front of everything) and the property access was 5.78% faster (I expected it to be slower, tbh)!
![image](https://user-images.githubusercontent.com/17912877/114499270-96cb1700-9bda-11eb-9bda-c0a11afcc8ee.png)

So yeah, I think if we could encapsulate the parser and change all the code to use `this.` then the hooks could easily pass `this` or similar. There were 2 properties that would be at the stack level, rather than the global parsing level though, `node`, and `nodes` but maybe we could get clever with that.

I'm fine with using ES6 classes (especially if we're using rollup to build CJS and ESM). That could easily be its own PR, before any other work is done.

Reply:
Oh wow, thank you so much for doing this, that's very illuminating!

Would you be ok with me doing the rewrite to use the ES6 class or do you want to do it?

Reply:
Totally feel free! (and if you feel like updating the build system/folder layout too, that would help ;-) )

Reply:
@LeaVerou - I just created #140 - hopefully it's okay? Or if you still want to do it (or do it differently) I'm not offended :-)